tools/base/fpgainfo/errors.c

```diff
@@ -0,0 +1,477 @@
+				size = PCIE1_ERROR_COUNT;
+				error_string = PCIE1_ERROR;
+			}
+
+			for (j = 0; (j < size) && (NULL != error_string); j++) {
+				if (FPGA_BIT_IS_SET(error_value, j)) {
+					printf("\t %s \n", error_string[j]);
+				}
+			}
+		}
+	} else if (((VERB_ALL == errors_config.which)
+		    || (VERB_PORT == errors_config.which))
+		   && (FPGA_ACCELERATOR == objtype)) {
+		fpgainfo_print_common("//****** PORT ERRORS ******//", props);
+
+		for (i = 0; i < (int)num_errors; i++) {
+			uint64_t error_value = 0;
+			res = fpgaReadError(token, i, &error_value);
+			fpgainfo_print_err("reading error for PORT", res);
+
+			printf("%-29s : 0x%" PRIX64 "\n", errinfos[i].name,
+			       error_value);
+
+			if (!strcmp(errinfos[i].name, "Errors")) {
+				size = PORT_ERROR_COUNT;
+				error_string = PORT_ERROR;
+			} else if (!strcmp(errinfos[i].name,
+					   "First Malformed Req")) {
+				size = 0;
+				error_string = NULL;
+			} else if (!strcmp(errinfos[i].name, "First Error")) {
+				size = 0;
+				error_string = NULL;
+			}
+
+			for (j = 0; (j < size) && (NULL != error_string); j++) {
+				if (FPGA_BIT_IS_SET(error_value, j)) {
+					printf("\t %s \n", error_string[j]);
+				}
+			}
+		}
+	}
+}
+
+fpga_result errors_command(fpga_token *tokens, int num_tokens, int argc,
+			   char *argv[])
+{
+	(void)argc;
+	(void)argv;
+	fpga_result res = FPGA_OK;
+	fpga_properties props;
+	struct fpga_error_info *errinfos = NULL;
+
+	if (errors_config.help_only) {
+		return res;
+	}
+
+	int i = 0;
+	for (i = 0; i < num_tokens; ++i) {
+		uint32_t num_errors;
+
+		res = fpgaGetProperties(tokens[i], &props);
+                if (res == FPGA_OK) {
+                        res = fpgaPropertiesGetNumErrors(props, &num_errors);
+                        fpgainfo_print_err("reading errors from properties", res);
+
+                        if (0 != num_errors) {
+                                int j;
+                                errinfos = (struct fpga_error_info *)calloc(
+                                        num_errors, sizeof(*errinfos));
+                                if (NULL == errinfos) {
+                                        fprintf(stderr, "Error allocating memory");
+                                        goto destroy_and_free;
+                                }
+
+                                for (j = 0; j < (int)num_errors; j++) {
+                                        res = fpgaGetErrorInfo(tokens[i], j,
+                                                               &errinfos[j]);
+                                        fpgainfo_print_err(
+                                                "reading error info structure", res);
+                                        (void)replace_chars(errinfos[j].name, '_', ' ');
+                                        (void)upcase_pci(errinfos[j].name);
+                                        (void)upcase_first(errinfos[j].name);
+                                }
```

Comment:
These three functions should probably return void if we're doing the mutation to the input string.